fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+        """
+        For each layer calculates the scaling factor for preceding layers' grad inputs
+        and current layers' grad outputs. These values are used to register a full backward
+        hook. The handle returned from registering the backward hook is appended to a list
+        of handles. New hooks are created and registered at every step and a new list of
+        handles is created. The handles are flushed out in the unscale function.
+        """
+        if self._apply_layerwise_scaling:
+            for idx in range(len(self.layer_info)):
+                elt = self.layer_info[idx]
+                layer_name, layer_type = elt.layer_name, elt.layer_type
+
+                inputs_multiplier = 1.0
+                if idx > 0:
+                    inputs_multiplier = self.layer_info[idx - 1].scaling_factor
+
+                outputs_multiplier = 1.0 / elt.scaling_factor
+                helper = GradientHelper(layer_name, inputs_multiplier, outputs_multiplier)
+                layer_handle = layer_type.register_full_backward_hook(helper.scale_gradients)
+                self._handles.append(layer_handle)
+                logging.debug("name = %s \t scale = %s" % (layer_name, elt.scaling_factor))
+
+    def unscale(self) -> None:
+        """
+        For each layer, check if any of the layers' parameters contain an inf/nan.
+        If there are no inf/nan in the gradient, then gradient of that layer is
+        unscaled by the reciprocal of the scaling factor for that layer.
+        Finally, all handles recorded while registering the hooks are deleted.
+        """
+        if self._apply_layerwise_scaling:
+            for elt in self.layer_info:
+                if not elt.found_inf_or_nan:
+                    for param_name, param in elt.layer_type.named_parameters():
+                        if hasattr(param, "grad"):
+                            logging.debug(
+                                "%s scaling down %s by %s" % (elt.layer_name, param_name, 1.0 / elt.scaling_factor)
+                            )
+                            param.grad = torch.mul(param.grad, 1.0 / elt.scaling_factor)
+
+            while len(self._handles) > 0:
+                elt = self._handles.pop()
+                elt.remove()
+
+    def _check_for_inf_or_nan(self) -> None:
+        """
+        For each layer, check if any of the parameters with a gradient attribute
+        contain an inf/nan. If any of the parameters' gradient contain an inf/nan,
+        then that layers' found_inf_or_nan attribute is set to True and all
+        remaining parameters for that layer are skipped.
+        """
+        for elt in self.layer_info:
+            elt.found_inf_or_nan = False
+            for _, param in elt.layer_type.named_parameters():
+                if hasattr(param, "grad") and param.grad is not None:
+                    if torch.isinf(param.grad).any().item() is True or torch.isnan(param.grad).any().item() is True:  # type: ignore
```

Comment:
do we need "is True" for these statements?

Reply:
They are not needed but if I remove them it's not clear what the expected returned type is. If it's 1 or "I wonder what the return type is"? Both would return True.

Reply:
The reason I mentioned this is that it can shorten the line since you can just check for the implicit True condition.

Reply:
I noticed the long line length as well but then I let the linter (black) decide on how to format it.

Reply:
So we set the line length to 120 here https://github.com/facebookresearch/fairscale/blob/562542478a145fae410cb23ffd46a8d7c2497abb/pyproject.toml#L8-L9
TIL: comment is not considered part of line length by black

Reply:
This is not a comment and checking for is True is redundant right? should we remove it?

Reply:
removed True